internal/shared/util/image/pull.go

```diff
@@ -0,0 +1,268 @@
+	//////////////////////////////////////////////////////
+	canonicalRef, err := resolveCanonicalRef(ctx, dockerImgRef, srcCtx)
+	if err != nil {
+		return nil, nil, time.Time{}, err
+	}
+	l = l.WithValues("digest", canonicalRef.Digest().String())
+
+	///////////////////////////////////////////////////////
+	//
+	// Check if the cache has already applied the
+	// canonical keep. If so, we're done.
+	//
+	///////////////////////////////////////////////////////
+	fsys, modTime, err := cache.Fetch(ctx, ownerID, canonicalRef)
+	if err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error checking cache for existing content: %w", err)
+	}
+	if fsys != nil {
+		return fsys, canonicalRef, modTime, nil
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Create an OCI layout reference for the destination,
+	// where we will temporarily store the image in order
+	// to unpack it.
+	//
+	// We use the OCI layout as a temporary storage because
+	// copy.Image can concurrently pull all the layers.
+	//
+	//////////////////////////////////////////////////////
+	layoutDir, err := os.MkdirTemp("", "oci-layout-*")
+	if err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error creating temporary directory: %w", err)
+	}
+	defer func() {
+		if err := os.RemoveAll(layoutDir); err != nil {
+			l.Error(err, "error removing temporary OCI layout directory")
+		}
+	}()
+
+	layoutImgRef, err := layout.NewReference(layoutDir, canonicalRef.String())
+	if err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error creating reference: %w", err)
+	}
+
+	//////////////////////////////////////////////////////
+	//
+	// Load an image signature policy and build
+	// a policy context for the image pull.
+	//
+	//////////////////////////////////////////////////////
+	policyContext, err := loadPolicyContext(srcCtx, l)
+	if err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error loading policy context: %w", err)
+	}
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			l.Error(err, "error destroying policy context")
+		}
+	}()
+
+	//////////////////////////////////////////////////////
+	//
+	// Pull the image from the source to the destination
+	//
+	//////////////////////////////////////////////////////
+	if _, err := copy.Image(ctx, policyContext, layoutImgRef, dockerImgRef, &copy.Options{
+		SourceCtx: srcCtx,
+		// We use the OCI layout as a temporary storage and
+		// pushing signatures for OCI images is not supported
+		// so we remove the source signatures when copying.
+		// Signature validation will still be performed
+		// accordingly to a provided policy context.
+		RemoveSignatures: true,
+	}); err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error copying image: %w", err)
+	}
+	l.Info("pulled image")
+
+	fsys, modTime, err = p.applyImage(ctx, ownerID, dockerRef, canonicalRef, layoutImgRef, cache, srcCtx)
+	if err != nil {
+		return nil, nil, time.Time{}, fmt.Errorf("error applying image: %w", err)
+	}
```

Comment:
This seems like another stage, where a large comment might be useful; the old code indicated:
```
        //////////////////////////////////////////////////////
        //
        // Mount the image we just pulled
        //
        //////////////////////////////////////////////////////